### PR TITLE
User mappings

### PR DESCRIPTION
**Pull request by gforcada**
This pull request only updates 2 out of 5 reports to take into account the user mappings.

### CODE REVIEW

File: z3c/dependencychecker/db.py [hunk 1]
```diff
@@ -275,3 +292,32 @@ def _get_test_extra(self):
...
+        """Remove dotted names that are metapackages (if they are used)"""
+        if reference is None:
+            reference = self._requirements
+
+        result = []
+        for single_import in dotted_names:
+            provided_package = False
+            for provided in self.reverse_metapackages:
+                if single_import in provided:
+                    provided_package = provided
```
Comment (reinout) [inline z3c/dependencychecker/db.py]: It is more "requirement that provides this package" than "provided package", I think.

So: is `providing_requirement` better?

File: z3c/dependencychecker/tests/test_package_overrides.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+two = "5.6.7.8"
+"""
+EMPTY_DEPENDENCY_TABLE_NESTED = """[tool]
+    [dependencycheker]
+"""
+EMPTY_DEPENDENCY_TABLE_MERGED = """[tool.dependencychecker]"""
+ONE_METAPACKAGE = """[tool.dependencychecker]
+Zope2 = ["Products.Five", ]
+"""
+MORE_METAPACKAGES = """[tool.dependencychecker]
```
Comment (reinout) [inline z3c/dependencychecker/tests/test_package_overrides.py]: Do we only want metapackages under this key? No other configuration (`verbose = True`, for instance)?

File: z3c/dependencychecker/db.py [hunk 3]
```diff
@@ -54,6 +56,15 @@ def add_imports(self, imports):
             logger.debug('    Import found: %s', single_import.name)
             self.imports_used.append(single_import)
 
+    def add_metapackage(self, metapackage, provided):
+        metapackage_obj = DottedName(metapackage)
+        provided_objects = [DottedName(name) for name in provided]
```
[the remark this answers is not in the document]
Comment (gforcada) [reply z3c/dependencychecker/db.py]: Fixed.

Yeah, maybe we could add a `Requirement` as a subclass of a `DottedName`, but I'm not sure if the equality and comparison will keep as they are... I'm a bit hesitant on this bit, although I find myself thinking the same every now and then, but somehow I manage to find a way out without breaking them into two.

File: z3c/dependencychecker/db.py [hunk 4]
```diff
@@ -19,6 +19,8 @@ def __init__(self):
         self._requirements = set()
         self._extras_requirements = {}
         self.imports_used = []
+        self.metapackages = {}
```
[the remark this answers is not in the document]
Comment (gforcada) [reply z3c/dependencychecker/db.py]: As we don't have a difference between the two of them (packages and requirements), should we call it right right away `user_mapping` and `reverse_user_mapping` ?

File: z3c/dependencychecker/db.py (hunk 1 above)
Comment (gforcada) [reply z3c/dependencychecker/db.py]: I reworked this bit as I was not really happy with the code, would you mind having another look?

File: z3c/dependencychecker/tests/test_package_overrides.py (hunk 2 above)
Comment (gforcada) [reply z3c/dependencychecker/tests/test_package_overrides.py]: Mostly to keep it simple. I expanded the code a little bit to take care only about listings, so for options and other things we could have, yet another subtable, i.e. `[tool.dependencychecker.options]`.

Sounds that good?

As soon as we introduce other options/configuration we can revisit the approach.

**Comment by gforcada** [comment]
@reinout thanks for the feedback, I did improve it a bit (I hope!).

Sorry for the delay, I was a bit frustrated with how the code ended up, so I needed it to rest for a bit to be able to have a fresh look at it.

Let me know, whenever you have time, if the code looks better now.

File: z3c/dependencychecker/db.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (reinout) [reply z3c/dependencychecker/db.py]: `user_mapping` sounds fine. Or perhaps just `mapping`: in that way we could add mappings of our own.

[the remark this answers is not in the document]
Comment (reinout) [reply z3c/dependencychecker/db.py]: Yeah, I really like the generic term "mapping" or "usermapping" or "packagemapping" or so, as it is more than just the real "meta mackages".

**Comment by gforcada** [comment]
@reinout I updated the code to use `user_mappings` everywhere and remove inconsistencies (on tests I was using `overrides`).

I also took the time to add a changelog entry and document the user mappings on the README file.

Let me know if it looks fine now.

**Comment by gforcada** [comment]
@reinout friendly ping :bell:  :slightly_smiling_face:

**Comment by reinout** [comment]
Oh, oopsie. Totally swamped with work. And this weekend I've got two family birthdays.
Perhaps @mauritsvanrees can take a very quick look?
Otherwise I'll look at it early next week :-)
Feel very free to ping me if I haven't done anything next week.

**Comment by reinout** [comment]
"User mappings" and "reverse user mappings" are perfect names, thanks!
Looks great, I'll release it right away.

**Comment by reinout** [comment]
I've released 2.1.1 (2.1 failed due to a recent bin/python upgrade on my homebrew install).

Thanks again for the work!

**Pull request merged**